quantum/action_tapping.c

```diff
@@ -708,6 +734,110 @@ void waiting_buffer_scan_tap(void) {
     }
 }
 
+#    ifdef SPECULATIVE_HOLD
+__attribute__((weak)) bool get_speculative_hold(uint16_t keycode, keyrecord_t *record) {
+    return (QK_MOD_TAP_GET_MODS(keycode) & (MOD_LALT | MOD_LGUI)) == 0;
+}
+
+static void debug_speculative_keys(void) {
+    ac_dprintf("mods = { ");
+    for (int8_t i = 0; i < num_speculative_keys; ++i) {
+        ac_dprintf("%02X ", speculative_keys[i].mods);
+    }
+    ac_dprintf("}, keys = { ");
+    for (int8_t i = 0; i < num_speculative_keys; ++i) {
+        ac_dprintf("%02X%02X ", speculative_keys[i].key.row, speculative_keys[i].key.col);
+    }
+    ac_dprintf("}\n");
+}
+
+static void process_speculative_press(keyrecord_t *record) {
+    if (num_speculative_keys >= SPECULATIVE_KEYS_SIZE) { // Overflow!
+        ac_dprintf("SPECULATIVE KEYS OVERFLOW: IGNORING EVENT\n");
+        return;
+    }
+
+    const uint16_t keycode = get_record_keycode(record, true);
+    if (!IS_QK_MOD_TAP(keycode)) {
+        return;
+    }
+
+    uint8_t mods = QK_MOD_TAP_GET_MODS(keycode);
+    if ((mods & 0x10) != 0) { // Unpack 5-bit mods to 8-bit representation.
+        mods <<= 4;
+    }
+
+    if ((~(get_mods() | get_speculative_mods()) & mods) != 0 && get_speculative_hold(keycode, record)) {
+        add_speculative_mods(mods);
+        speculative_keys[num_speculative_keys] = (speculative_key_t){
+            .key  = record->event.key,
```

Comment:
I'm still doing an initial pass so am not familiar with the code yet - but would storing the record->event.pos not be better? I imagine we only need this to release early (before the actual speculative-ness is collapsed and key events are sent) in which case checking the physical location seems ideal.

Reply:
I just fumbled some keys and was left with control stuck down. I can't manage to reproduce it, but it involved a `LCTL(KC_LEFT)` under a layer tap. Not sure if I hit my Retro Shifted `MT(MOD_LCTL, KC_SCLN)` on the main layer as well. This keymap has never stuck keys with a few years of use, though I did just flash those years' worth of QMK updates alongside this PR.

I'd definitely feel safer using `event->pos` if possible :slightly_smiling_face: just less chance for weird cases.

Reply:
Thanks for the report and heads up, I'll see if I can repro that!

"`.key`" is perhaps a misnomer, it's a `keypos_t`, not a 16-bit keycode. I very much agree that keys need to be tracked based on matrix position, due to the layer ambiguity as described in the other comment.